internal/mode/static/state/graph/route_common.go

```diff
@@ -0,0 +1,567 @@
+				NamespacedName: client.ObjectKeyFromObject(route),
+				RouteType:      RouteTypeHTTP,
+			}
+			routes[rk] = r
+		}
+	}
+
+	for _, route := range grpcRoutes {
+		r := buildGRPCRoute(validator, route, gatewayNsNames)
+		if r != nil {
+			rk := RouteKey{
+				NamespacedName: client.ObjectKeyFromObject(route),
+				RouteType:      RouteTypeGRPC,
+			}
+			routes[rk] = r
+		}
+	}
+
+	return routes
+}
+
+func buildSectionNameRefs(
+	parentRefs []v1.ParentReference,
+	routeNamespace string,
+	gatewayNsNames []types.NamespacedName,
+) ([]ParentRef, error) {
+	sectionNameRefs := make([]ParentRef, 0, len(parentRefs))
+
+	type key struct {
+		gwNsName    types.NamespacedName
+		sectionName string
+	}
+	uniqueSectionsPerGateway := make(map[key]struct{})
+
+	for i, p := range parentRefs {
+		gw, found := findGatewayForParentRef(p, routeNamespace, gatewayNsNames)
+		if !found {
+			continue
+		}
+
+		var sectionName string
+		if p.SectionName != nil {
+			sectionName = string(*p.SectionName)
+		}
+
+		k := key{
+			gwNsName:    gw,
+			sectionName: sectionName,
+		}
+
+		if _, exist := uniqueSectionsPerGateway[k]; exist {
+			return nil, fmt.Errorf("duplicate section name %q for Gateway %s", sectionName, gw.String())
+		}
+		uniqueSectionsPerGateway[k] = struct{}{}
+
+		sectionNameRefs = append(sectionNameRefs, ParentRef{
+			Idx:     i,
+			Gateway: gw,
+		})
+	}
+
+	return sectionNameRefs, nil
+}
+
+func findGatewayForParentRef(
+	ref v1.ParentReference,
+	routeNamespace string,
+	gatewayNsNames []types.NamespacedName,
+) (gwNsName types.NamespacedName, found bool) {
+	if ref.Kind != nil && *ref.Kind != "Gateway" {
+		return types.NamespacedName{}, false
+	}
+	if ref.Group != nil && *ref.Group != v1.GroupName {
+		return types.NamespacedName{}, false
+	}
+
+	// if the namespace is missing, assume the namespace of the HTTPRoute
+	ns := routeNamespace
+	if ref.Namespace != nil {
+		ns = string(*ref.Namespace)
+	}
+
+	for _, gw := range gatewayNsNames {
+		if gw.Namespace == ns && gw.Name == string(ref.Name) {
+			return gw, true
+		}
+	}
+
+	return types.NamespacedName{}, false
+}
+
+func bindRoutesToListeners(
+	routes map[RouteKey]*L7Route,
+	gw *Gateway,
+	namespaces map[types.NamespacedName]*apiv1.Namespace,
+) {
+	if gw == nil {
+		return
+	}
+
+	for _, r := range routes {
+		bindRouteToListeners(r, gw, namespaces)
+	}
+}
+
+func bindRouteToListeners(
+	route *L7Route,
+	gw *Gateway,
+	namespaces map[types.NamespacedName]*apiv1.Namespace,
+) {
+	if !route.Attachable {
+		return
+	}
+
+	bindRoutes := func(r *L7Route) {
+		for i := 0; i < len(r.ParentRefs); i++ {
+			attachment := &ParentRefAttachmentStatus{
+				AcceptedHostnames: make(map[string][]string),
+			}
+			ref := &r.ParentRefs[i]
+			ref.Attachment = attachment
+
+			routeRef := r.SrcParentRefs[ref.Idx]
+
+			path := field.NewPath("spec").Child("parentRefs").Index(ref.Idx)
+
+			attachableListeners, listenerExists := findAttachableListeners(
+				getSectionName(routeRef.SectionName),
```

Comment:
Can we add SectionName to the ParentRef and get rid of SrcParentRefs?

Reply:
That's a good point 🤦🏼‍♀️